std/utf.d

```diff
@@ -2420,6 +2469,167 @@ unittest // bugzilla 12923
     }());
 }
 
+/++
+    Returns whether the given range of characters is well-formed Unicode or not.
+
+    Params:
+        range = The range of charaters to validate.
+        index = The index (in code units) where the failed character was (or `0`
+                if the result is `true`).
+  +/
+pragma(inline, true) bool isValidUnicode(R)(R range)
+    if (isInputRange!R && isSomeChar!(ElementType!R) && !isInfinite!R)
+{
+    size_t failedIndex;
+    return _isValidUnicodeImplTrustWrap!false(range, failedIndex);
+}
+
+/// Ditto
+pragma(inline, true) bool isValidUnicode(R)(R range, out size_t failedIndex)
+    if (isInputRange!R && isSomeChar!(ElementType!R) && !isInfinite!R)
+{
+    return _isValidUnicodeImplTrustWrap!true(range, failedIndex);
+}
+
+pragma(inline, true)
+private bool _isValidUnicodeImplTrustWrap(bool getIndex, R)(R range, out size_t failedIndex) @trusted
```

Comment:
This is doing basically the same thing that `decode` does. The overload that operates on strings is `@trusted`. It has to be to be `@safe`, because `decodeImpl` uses `.ptr` and pointer arithmetic for strings. And you can't put `@trusted` on `decodeImpl`, because it operates on arbitrary ranges of characters, which could have `@system` range functions. So, the choice with `isValidUnicode` was either to create 4 overloads so that the string ones can be `@trusted` and the arbitrary ranges not or to have two overloads and create an internal wrapper which is overloaded such that the string overload is `@trusted` and the one for arbitrary ranges is not. We're stuck with slapping `@trusted` on the overloads that take strings one way or another regardless because of how `decodeImpl` uses pointer arithmetic for efficiency. Trying to wrap that in `decodeImpl` would be a nightmare and risk killing the efficiency gain, and overloading `decodeImpl` would result in duplicating the core Unicode decoding code, which seems like a really bad idea. `decodeImpl` is very carefully written to be efficient for strings while still working with arbitrary ranges and avoiding the need to duplicate the core logic.